crypto/fipsmodule/rand/entropy_pool_test.cc

```diff
@@ -0,0 +1,190 @@
+  //  * Consume zero bytes.
+  //  * Use NULL as argument for entropy pool.
+  //  * Use NULL as argument for output buffer.
+  EXPECT_FALSE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, ENTROPY_POOL_SIZE+1));
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+  EXPECT_FALSE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, 0));
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+  EXPECT_FALSE(RAND_entropy_pool_get(NULL, entropy_buffer, ENTROPY_POOL_SIZE));
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+  EXPECT_FALSE(RAND_entropy_pool_get(&entropy_pool, NULL, ENTROPY_POOL_SIZE));
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+
+  // Modify internal state to capture validations
+  OPENSSL_cleanse(entropy_buffer, ENTROPY_POOL_SIZE);
+  RAND_entropy_pool_add(&entropy_pool, fake_entropy);
+  entropy_pool.capacity = ENTROPY_POOL_SIZE + 1;
+  EXPECT_FALSE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, ENTROPY_POOL_SIZE));
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+  entropy_pool.capacity = ENTROPY_POOL_SIZE;
+  EXPECT_TRUE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, ENTROPY_POOL_SIZE));
+
+  OPENSSL_cleanse(entropy_buffer, ENTROPY_POOL_SIZE);
+  RAND_entropy_pool_add(&entropy_pool, fake_entropy);
+  entropy_pool.valid_available = ENTROPY_POOL_SIZE + 1;
+  EXPECT_FALSE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, ENTROPY_POOL_SIZE));
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+  entropy_pool.valid_available = ENTROPY_POOL_SIZE;
+  EXPECT_TRUE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, ENTROPY_POOL_SIZE));
+
+  OPENSSL_cleanse(entropy_buffer, ENTROPY_POOL_SIZE);
+  RAND_entropy_pool_add(&entropy_pool, fake_entropy);
+  entropy_pool.index_read = ENTROPY_POOL_SIZE + 1;
+  EXPECT_FALSE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, ENTROPY_POOL_SIZE));
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+  entropy_pool.index_read = 0;
+  EXPECT_TRUE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, ENTROPY_POOL_SIZE));
+}
+
+TEST(EntropyPool, EntirePoolRefreshedAtAddTime) {
+  uint8_t entropy_buffer[ENTROPY_POOL_SIZE] = {0};
+  struct entropy_pool entropy_pool;
+  uint8_t fake_entropy_A[ENTROPY_POOL_SIZE] = {0};
+  fill_fake_entropy(fake_entropy_A, 'A');
+  uint8_t fake_entropy_B[ENTROPY_POOL_SIZE] = {0};
+  fill_fake_entropy(fake_entropy_B, 'B');
+
+  // Even though there is still valid available entropy in the pool, the entire
+  // pool must be refreshed when adding fresh entropy.
+  RAND_entropy_pool_init(&entropy_pool);
+  RAND_entropy_pool_add(&entropy_pool, fake_entropy_A);
+  EXPECT_TRUE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, entropy_pool_size_halfed));
+  EXPECT_GT(entropy_pool.valid_available, (size_t) 0);
```

Comment:
Can this check for a specific value?

Reply:
You'll need to compute it, but then also check that the computed value is strictly larger than 0 to do the test I want to do. So, it all collapse to this test.